File: app/src/main/java/net/squanchy/eventdetails/widget/EventDetailsCoordinatorLayout.java
package net.squanchy.eventdetails.widget;

import android.content.Context;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.FloatingActionButton;
import android.util.AttributeSet;

import net.squanchy.R;
import net.squanchy.schedule.domain.view.Event;

import net.squanchy.schedule.domain.view.Event.Type;
import net.squanchy.support.widget.SpeakerView;

public class EventDetailsCoordinatorLayout extends CoordinatorLayout {

    private FloatingActionButton floatingActionButton;
    private EventDetailsHeaderLayout headerLayout;
    private EventDetailsLayout detailsLayout;

    public EventDetailsCoordinatorLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public EventDetailsCoordinatorLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();

        floatingActionButton = (FloatingActionButton) findViewById(R.id.favorite_fab);
        headerLayout = (EventDetailsHeaderLayout) findViewById(R.id.event_details_header);
        detailsLayout = (EventDetailsLayout) findViewById(R.id.event_details);
    }

    public void updateWith(Event event, OnEventDetailsClickListener listener) {
        headerLayout.updateWith(event, listener);
        detailsLayout.updateWith(event);

        if (canBeFavorited(event)) {
            floatingActionButton.setImageResource(
                    event.favorited() ?
                            R.drawable.ic_favorite_filled :
                            R.drawable.ic_favorite_empty
            );
            floatingActionButton.setOnClickListener(v -> listener.onFavoriteClick());
            floatingActionButton.setVisibility(VISIBLE);
        } else {
            floatingActionButton.setVisibility(GONE);
        }
    }

    private boolean canBeFavorited(Event event) {
        Event.Type type = event.type();
        return type == Type.TALK || type == Type.KEYNOTE;
    }

    public interface OnEventDetailsClickListener extends OnFavoriteClickListener, SpeakerView.OnSpeakerClickListener {

    }

    public interface OnFavoriteClickListener {
        void onFavoriteClick();
    }
}

File: app/src/main/java/net/squanchy/eventdetails/widget/EventDetailsCoordinatorLayout.kt
package net.squanchy.eventdetails.widget

import android.content.Context
import android.util.AttributeSet
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.view.isVisible
import kotlinx.android.synthetic.main.activity_event_details.view.*
import net.squanchy.schedule.domain.view.Event
import net.squanchy.schedule.domain.view.Event.Type
import net.squanchy.support.widget.SpeakerView

class EventDetailsCoordinatorLayout @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet,
    defStyleAttr: Int = 0
) : CoordinatorLayout(context, attrs, defStyleAttr) {

    internal fun updateWith(event: Event, listener: OnEventDetailsClickListener) {
        eventDetailsHeaderLayout.updateWith(event, listener)
        eventDetailsLayout.updateWith(event)

        if (event.canBeFavorited) {
            // Updates the FAB image state to trigger an AVD animation.
            val stateSet = intArrayOf(android.R.attr.state_checked * if (event.favorite) 1 else -1)
            favoriteFab.setImageState(stateSet, true)
            favoriteFab.setOnClickListener { listener.onFavoriteClick() }
            favoriteFab.isVisible = true
        } else {
            favoriteFab.isVisible = false
        }
    }

    private val Event.canBeFavorited
        get() = type == Type.TALK || type == Type.KEYNOTE || type == Type.WORKSHOP

    internal interface OnEventDetailsClickListener : OnFavoriteClickListener, SpeakerView.OnSpeakerClickListener

    internal interface OnFavoriteClickListener {

        fun onFavoriteClick()
    }
}
